### PR TITLE
LF: Structure package loading errors

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
part of #9974.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala [hunk 1]
```diff
@@ -428,31 +435,17 @@ class Engine(val config: EngineConfig = new EngineConfig(LanguageVersion.StableV
...
-        Either.cond(
-          missingDeps.isEmpty,
-          (),
-          Error.Package.Generic(
-            s"The set of packages ${pkgIds.mkString("{'", "', '", "'}")} is not self consistent, the missing dependencies are ${missingDeps
-              .mkString("{'", "', '", "'}")}."
-          ),
-        )
-      }
+      unknownPackages = pkgIds.filterNot(pkgs.isDefinedAt)
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:445]: Looking at the usages of this function I’m not entirely sure the API here makes sense. We pass in a `Set[PackageId]` and a `Map[PackageId, Package]` and then assert that the set contains the key set of the map. We should be able to just accept the map only, eliminate this error and afaict all callsites would work well with that. Probably something for a separate PR though.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala [hunk 2]
```diff
@@ -54,7 +54,7 @@ private[engine] final class Preprocessor(compiledPackages: MutableCompiledPackag
                 )
               } yield r
             case None =>
-              ResultError(Error.Package.Generic(s"Couldn't find package $pkgId"))
+              ResultError(Error.Package.MissingPackage(pkgId))
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala:57]: I’m a bit worried that there seem no tests for any of the other errors.  Am I missing something? At least self consistency seems worth testing somewhere™.

**Review by cocreature — APPROVED** [review]
LGTM with a few comments, thanks!

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:445]: https://github.com/digital-asset/daml/pull/9986

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala [hunk 3]
```diff
@@ -246,7 +247,12 @@ class Engine(val config: EngineConfig = new EngineConfig(LanguageVersion.StableV
         case speedy.Compiler.PackageNotFound(_) =>
           handleMissingDependencies.flatMap(_ => start)
         case speedy.Compiler.CompilationError(error) =>
-          ResultError(Error.Package.Generic(s"CompilationError: $error"))
+          ResultError(
+            Error.Package.Internal(
+              methodName,
+              s"CompilationError: $error",
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:253]: In fact, this is not a `Package.Error`, I move it to `Preprocessing.Error`.
Good catch,

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala:57]: I added tests and spotted/fixed a bug in 2ff96785562ac41f2793aff38ec889b6075595a0

**Pull request merged**